build-system/pr-check.js

```diff
@@ -0,0 +1,190 @@
+ * @param {string} cmd
+ */
+function execOrDie(cmd) {
+  console /*OK*/.log(`\npr-check.js: ${cmd}\n`);
+  const p =
+      child_process.spawnSync('/bin/sh', ['-c', cmd], {'stdio': 'inherit'});
+  if (p.status != 0) {
+    console /*OK*/.error(
+        `\npr-check.js - exiting due to failing command: ${cmd}\n`);
+    process.exit(p.status)
+  }
+}
+
+/**
+ * For a provided commit range identifiying a pull request (PR),
+ * yields the list of files.
+ * @param {string} travisCommitRange
+ * @return {!Array<string>}
+ */
+function filesInPr(travisCommitRange) {
+  return exec(`git diff --name-only ${travisCommitRange}`);
+}
+
+/**
+ * Determines whether the given file belongs to the Validator webui,
+ * that is, the 'VALIDATOR_WEBUI' target.
+ * @param {string} filePath
+ * @return {boolean}
+ */
+function isValidatorWebuiFile(filePath) {
+  return hasPrefix(filePath, 'validator/webui');
+}
+
+/**
+ * Determines whether the given file belongs to the validator,
+ * that is, the 'VALIDATOR' target. This assumes (but does not
+ * check) that the file is not part of 'VALIDATOR_WEBUI'.
+ * @param {string} filePath
+ * @return {boolean}
+ */
+function isValidatorFile(filePath) {
+  if (hasPrefix(filePath, 'validator/')) return true;
+  if (!hasSuffix(path.dirname(filePath), '0.1') &&
+      !hasSuffix(path.dirname(filePath), 'test'))
+    return false;
+  const name = path.basename(filePath);
+  return hasPrefix(name, 'validator-') &&
+      (hasSuffix(name, '.out') || hasSuffix(name, '.html') ||
+       hasSuffix('.protoascii'));
+}
+
+/**
+ * Determines the targets that will be executed by the main method of
+ * this script. The order within this function matters.
+ * @param {!Array<string>} filePaths
+ * @returns {!Set<string>}
+ */
+function determineBuildTargets(filePaths) {
+  const targetSet = new Set();
+  for (p of filePaths) {
+    if (isValidatorWebuiFile(p)) {
+      targetSet.add('VALIDATOR_WEBUI');
+    } else if (isValidatorFile(p)) {
+      targetSet.add('VALIDATOR');
+    } else {
+      targetSet.add('RUNTIME');
+    }
+  }
+  return targetSet;
+}
+
+/**
+ * The main method for the script execution which much like a C main function
+ * receives the command line arguments and returns an exit status.
+ * @param {!Array<string>} argv
+ * @returns {number}
+ */
+function main(argv) {
+  if (argv.length <= 2) {
+    console /*OK*/.error(`Usage: ${__filename} TRAVIS_COMMIT_RANGE`);
+    return -1;
+  }
+  const travisCommitRange = argv[2];
+  const buildTargets = determineBuildTargets(filesInPr(travisCommitRange));
+
+  const sortedBuildTargets = [];
+  for (const t of buildTargets) {
+    sortedBuildTargets.push(t);
+  }
+  sortedBuildTargets.sort();
+
+  console /*OK*/.log(
+      '\npr-check.js: detected build targets: ' +
+      sortedBuildTargets.join(', ') + '\n');
+
+  execOrDie('npm run ava');
+  execOrDie('gulp lint');
+  execOrDie('gulp build --css-only');
```

Comment:
I'm not sure which of this if any would run over the validator. Probably not. So I've moved it into RUNTIME. We probably do want to run gulp presubmit though, so I left that part as is.